<a href="https://colab.research.google.com/github/dhiruvivek/Book-Recommendation/blob/main/Books_Dataset_Treatment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Book Recommendation System**    - 



##### **Project Type**    - Unsupervised Machine Learning Capstone Project 
##### **Contribution**    - Individual
##### **Name -** Vivek Tripathi


# **GitHub Link -**

https://github.com/dhiruvivek/Book-Recommendation

# **Project Summary -**
A book recommendation system is a system that suggests books to users based on their interests and preferences. These systems are commonly used by online bookstores, libraries, and reading apps to help users discover new books and authors. There are several approaches to building a book recommendation system. One common approach is to use collaborative filtering, which involves analyzing the reading patterns and preferences of a group of users and making recommendations based on what similar users have enjoyed. Another approach is to use content-based filtering, which involves analyzing the characteristics of a book (such as its genre, author, and subject matter) and making recommendations based on how closely they match the user's interests.

In this project, building a book recommendation system using one or more of these approaches.need to gather a dataset of books and user ratings, and then use machine learning algorithms to build a model that can make recommendations. You may also need to design a user interface for your system, so that users can easily interact with it and receive recommendations. Overall, this project will involve a combination of data gathering, data analysis, and development skills, and will give the opportunity to apply knowledge of machine learning and recommendation systems in a practical context.

# **Objective**


A book recommendation system is a type of recommendation system where we have to recommend similar books to the reader based on his interest. The books recommendation system is used by online websites which provide ebooks like google play books, open library, good Read’s, etc.

## **Understanding the Business Problem**

Businesses may have the following expectations from a book recommendation system:

**Increased sales**: One of the main goals of a book recommendation system is to increase sales by suggesting books that users are likely to purchase. This can be achieved by making personalized recommendations that match the user's interests and preferences.

**Customer satisfaction**: A good recommendation system should be able to provide users with a satisfying and enjoyable experience. This can be achieved by making relevant and accurate recommendations and by providing a user-friendly interface.

**User engagement**: A recommendation system can help to keep users engaged with a business's platform by providing a continuous stream of recommendations and new content. This can lead to increased customer loyalty and a longer user lifespan.

**Improved user experience**: A recommendation system can help businesses to improve the overall user experience by making it easier for users to discover new books and authors that they may enjoy.

**Competitive advantage**: A well-designed recommendation system can provide businesses with a competitive advantage over their rivals, as it can help to attract and retain customers.

Data-driven decision making: A recommendation system can provide businesses with insights into user preferences and reading habits, which can be used to inform marketing and content strategy decisions.

## **Importing Libraries**

In [ ]:
import numpy as np
import pandas as pd
import scipy
import math
import random
import sklearn
from nltk.corpus import stopwords
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from scipy.sparse.linalg import svds
import matplotlib.pyplot as plt
import seaborn as sns
import sys
import plotly.express as px
import plotly.graph_objects as go
import warnings
warnings.filterwarnings('ignore')
     

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
books=pd.read_csv("/content/drive/MyDrive/Book Recommendation SYstem/Books.csv")
users=pd.read_csv("/content/drive/MyDrive/Book Recommendation SYstem/Users.csv")
ratings=pd.read_csv("/content/drive/MyDrive/Book Recommendation SYstem/Ratings.csv")

In [ ]:
#Top 5 rows of Books data
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

**Data set Column information for Books data set**


*   ISBN (It is the unique number for different books)

*   Book-Title (It is the title of the books)

*   Book-Author (The author of the book)
*   Year-Of-Publication (It is the year when book was published)


*   Publisher (It is the name of the publisher)


*   Image-URL-S (It is URL of the books categorizing it as small)


*   Image-URL-M (It is URL of the books categorizing it as medium)
*   Image-URL-L (It is URL of the books categorizing it as large)













In [ ]:
# lets describe the Book dataset
books.describe()

ISBN      Book-Title      Book-Author  Year-Of-Publication  \
count       271360          271360           271359               271360   
unique      271360          242135           102023                  202   
top     0195153448  Selected Poems  Agatha Christie                 2002   
freq             1              27              632                13903   

        Publisher                                        Image-URL-S  \
count      271358                                             271360   
unique      16807                                             271044   
top     Harlequin  http://images.amazon.com/images/P/185326119X.0...   
freq         7535                                                  2   

                                              Image-URL-M  \
count                                              271360   
unique                                             271044   
top     http://images.amazon.com/images/P/185326119X.0...   
freq                                                    2   

                                              Image-URL-L  
count                                              271357  
unique                                             271041  
top     http://images.amazon.com/images/P/225307649X.0...  
freq                                                    2

In [ ]:
#Chcek the Dtype of Columns in Books data
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271359 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [ ]:
#lets check the unique values of columns of Books dataset
books.apply(lambda x: len(x.unique()))

ISBN                   271360
Book-Title             242135
Book-Author            102024
Year-Of-Publication       202
Publisher               16808
Image-URL-S            271044
Image-URL-M            271044
Image-URL-L            271042
dtype: int64

All the features has many unique values.And Year of publication has minimum unique values.

In [ ]:
# Checking for null values
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

#Lets check the unique values of different features of the books dataset.

In [ ]:

books['Book-Title'].unique()

array(['Classical Mythology', 'Clara Callan', 'Decision in Normandy', ...,
       'Lily Dale : The True Story of the Town that Talks to the Dead',
       "Republic (World's Classics)",
       "A Guided Tour of Rene Descartes' Meditations on First Philosophy with Complete Translations of the Meditations by Ronald Rubin"],
      dtype=object)

In [ ]:
books['Book-Author'].unique()

array(['Mark P. O. Morford', 'Richard Bruce Wright', "Carlo D'Este", ...,
       'David Biggs', 'Teri Sloat', 'Christopher  Biffle'], dtype=object)

In [ ]:
books['Image-URL-S'].unique()

array(['http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg',
       'http://images.amazon.com/images/P/0002005018.01.THUMBZZZ.jpg',
       'http://images.amazon.com/images/P/0060973129.01.THUMBZZZ.jpg',
       ...,
       'http://images.amazon.com/images/P/006008667X.01.THUMBZZZ.jpg',
       'http://images.amazon.com/images/P/0192126040.01.THUMBZZZ.jpg',
       'http://images.amazon.com/images/P/0767409752.01.THUMBZZZ.jpg'],
      dtype=object)

In [ ]:
books.Publisher.unique()

array(['Oxford University Press', 'HarperFlamingo Canada',
       'HarperPerennial', ..., 'Tempo', 'Life Works Books', 'Connaught'],
      dtype=object)

In [ ]:
books['Year-Of-Publication'].unique()

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

we can see that their are some values in Year-of-publication column like,'DK Publishing Inc','Gallimard','2024','2037','2030',2038, 2026,which could not be there in dataset so,we have to manipulate it before developing a recommender system.

In [ ]:
# lets check in how many rows are the year as DK Publishing inc
books.loc[books['Year-Of-Publication']=='DK Publishing Inc']

ISBN                                         Book-Title  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

       Book-Author Year-Of-Publication  \
209538        2000   DK Publishing Inc   
221678        2000   DK Publishing Inc   

                                                Publisher  \
209538  http://images.amazon.com/images/P/078946697X.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-S  \
209538  http://images.amazon.com/images/P/078946697X.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-M Image-URL-L  
209538  http://images.amazon.com/images/P/078946697X.0...         NaN  
221678  http://images.amazon.com/images/P/0789466953.0...         NaN

There are two rows in which the Year of Publication is as 'DK Publishing Inc'.We can see that the values are written at wrong places lets make it as correct place.

In [ ]:
# Lets correct the values of index number 209538 and 221678 
books.at[209538 ,'Publisher'] = 'DK Publishing Inc'
books.at[209538 ,'Year-Of-Publication'] = '2000'
books.at[209538 ,'Book-Title'] = 'DK Readers: Creating the X-Men, How It All Began (Level 4: Proficient Readers)'
books.at[209538 ,'Book-Author'] = 'Michael Teitelbaum'

books.at[221678 ,'Publisher'] = 'DK Publishing Inc'
books.at[221678 ,'Year-Of-Publication'] = '2000'
books.at[221687 ,'Book-Title'] = 'DK Readers: Creating the X-Men, How Comic Books Come to Life (Level 4: Proficient Readers)'
books.at[221678 ,'Book-Author'] = 'James Buckley'

In [ ]:
# now lets check values of index 209538
books.loc[209538]

ISBN                                                          078946697X
Book-Title             DK Readers: Creating the X-Men, How It All Beg...
Book-Author                                           Michael Teitelbaum
Year-Of-Publication                                                 2000
Publisher                                              DK Publishing Inc
Image-URL-S            http://images.amazon.com/images/P/078946697X.0...
Image-URL-M            http://images.amazon.com/images/P/078946697X.0...
Image-URL-L                                                          NaN
Name: 209538, dtype: object

In [ ]:
# now lets check values of index 221678
books.loc[221678]

ISBN                                                          0789466953
Book-Title             DK Readers: Creating the X-Men, How Comic Book...
Book-Author                                                James Buckley
Year-Of-Publication                                                 2000
Publisher                                              DK Publishing Inc
Image-URL-S            http://images.amazon.com/images/P/0789466953.0...
Image-URL-M            http://images.amazon.com/images/P/0789466953.0...
Image-URL-L                                                          NaN
Name: 221678, dtype: object

In [ ]:
# Lets check in hiw many rows are the year of publicatiobn as gallimard
books.loc[books['Year-Of-Publication']=='Gallimard']

ISBN                                         Book-Title  \
220731  2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...   

       Book-Author Year-Of-Publication  \
220731        2003           Gallimard   

                                                Publisher  \
220731  http://images.amazon.com/images/P/2070426769.0...   

                                              Image-URL-S  \
220731  http://images.amazon.com/images/P/2070426769.0...   

                                              Image-URL-M Image-URL-L  
220731  http://images.amazon.com/images/P/2070426769.0...         NaN

only one row there is year of publication as Gallimard.We can see that the values are written at wrong places.So lets write at its correct place.

In [ ]:
# lets coorect the values at index 220731
books.at[220731,'Book-Author']='Jean-Marie Gustave Le'
books.at[220731,'Book-Title']='Peuple du cel,suivi de les Bergers'
books.at[220731,'Publisher']='Gallimard'
books.at[220731,'Year-Of-Publication']='2003'

In [ ]:
# lets check the values of index 220731
books.loc[220731]

ISBN                                                          2070426769
Book-Title                            Peuple du cel,suivi de les Bergers
Book-Author                                        Jean-Marie Gustave Le
Year-Of-Publication                                                 2003
Publisher                                                      Gallimard
Image-URL-S            http://images.amazon.com/images/P/2070426769.0...
Image-URL-M            http://images.amazon.com/images/P/2070426769.0...
Image-URL-L                                                          NaN
Name: 220731, dtype: object

In [ ]:
#since we are living in 2023 so we have to remove values in which the year of publication is greater than 2023.
# Converting year of publication in Numbers
books['Year-Of-Publication'] = books['Year-Of-Publication'].astype(int)
print(sorted(books['Year-Of-Publication'].unique()))

[0, 1376, 1378, 1806, 1897, 1900, 1901, 1902, 1904, 1906, 1908, 1909, 1910, 1911, 1914, 1917, 1919, 1920, 1921, 1922, 1923, 1924, 1925, 1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936, 1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2008, 2010, 2011, 2012, 2020, 2021, 2024, 2026, 2030, 2037, 2038, 2050]


In [ ]:
# lets check how many values of year of publication are greater than 2022 years
books.loc[books['Year-Of-Publication']>2023].count()

ISBN                   13
Book-Title             13
Book-Author            13
Year-Of-Publication    13
Publisher              13
Image-URL-S            13
Image-URL-M            13
Image-URL-L            13
dtype: int64

In [ ]:
# lets check which are the year of publications which is greater than 2023
books.loc[books['Year-Of-Publication']>2023]

ISBN                                         Book-Title  \
37487   0671746103  MY TEACHER FRIED MY BRAINS (RACK SIZE) (MY TEA...   
55676   0671791990  MY TEACHER FLUNKED THE PLANET (RACK SIZE) (MY ...   
78168   0870449842                                   Crossing America   
80264   0140301690  Alice's Adventures in Wonderland and Through t...   
97826   0140201092      Outline of European Architecture (Pelican S.)   
116053  0394701658                       Three Plays of Eugene Oneill   
118294  3442436893        Das gro��e B�¶se- M�¤dchen- Lesebuch.   
192993  0870446924  Field Guide to the Birds of North America, 3rd...   
228173  0671266500       FOREST PEOPLE (Touchstone Books (Hardcover))   
240169  0684718022            In Our Time: Stories (Scribner Classic)   
246842  0380000059                                              CLOUT   
255409  068471809X                               To Have and Have Not   
260974  0671740989        FOOTBALL SUPER TEAMS : FOOTBALL SUPER TEAMS   

                        Book-Author  Year-Of-Publication  \
37487                       Coville                 2030   
55676                 Bruce Coville                 2030   
78168   National Geographic Society                 2030   
80264                 Lewis Carroll                 2050   
97826              Nikolaus Pevsner                 2050   
116053               Eugene O'Neill                 2038   
118294                  Kathy Lette                 2026   
192993  National Geographic Society                 2030   
228173            Colin M. Turnbull                 2030   
240169             Ernest Hemingway                 2030   
246842                   D. GIBBONS                 2024   
255409             Ernest Hemingway                 2037   
260974                  Bill Gutman                 2030   

                                         Publisher  \
37487                                      Aladdin   
55676                                      Aladdin   
78168                          National Geographic   
80264                                 Puffin Books   
97826                                  Penguin USA   
116053                           Vintage Books USA   
118294                                    Goldmann   
192993                         National Geographic   
228173                        Simon &amp; Schuster   
240169                               Collier Books   
246842                                        Avon   
255409                        Simon &amp; Schuster   
260974  Simon &amp; Schuster Children's Publishing   

                                              Image-URL-S  \
37487   http://images.amazon.com/images/P/0671746103.0...   
55676   http://images.amazon.com/images/P/0671791990.0...   
78168   http://images.amazon.com/images/P/0870449842.0...   
80264   http://images.amazon.com/images/P/0140301690.0...   
97826   http://images.amazon.com/images/P/0140201092.0...   
116053  http://images.amazon.com/images/P/0394701658.0...   
118294  http://images.amazon.com/images/P/3442436893.0...   
192993  http://images.amazon.com/images/P/0870446924.0...   
228173  http://images.amazon.com/images/P/0671266500.0...   
240169  http://images.amazon.com/images/P/0684718022.0...   
246842  http://images.amazon.com/images/P/0380000059.0...   
255409  http://images.amazon.com/images/P/068471809X.0...   
260974  http://images.amazon.com/images/P/0671740989.0...   

                                              Image-URL-M  \
37487   http://images.amazon.com/images/P/0671746103.0...   
55676   http://images.amazon.com/images/P/0671791990.0...   
78168   http://images.amazon.com/images/P/0870449842.0...   
80264   http://images.amazon.com/images/P/0140301690.0...   
97826   http://images.amazon.com/images/P/0140201092.0...   
116053  http://images.amazon.com/images/P/0394701658.0...   
118294  http://images.amazon.com/images/P/3442436893.0...   
192993  http://images.amazon.com/images/P/08704469

In [ ]:
# lets check how many values of year of publication are equal to zero
books.loc[books['Year-Of-Publication']==0].count()

ISBN                   4618
Book-Title             4618
Book-Author            4618
Year-Of-Publication    4618
Publisher              4618
Image-URL-S            4618
Image-URL-M            4618
Image-URL-L            4618
dtype: int64

In [ ]:
# lets see the Year of publication which is 0
books.loc[books['Year-Of-Publication']==0]

ISBN                                         Book-Title  \
176     3150000335                                   Kabale Und Liebe   
188     342311360X                            Die Liebe in Den Zelten   
288     0571197639                        Poisonwood Bible Edition Uk   
351     3596214629       Herr Der Fliegen (Fiction, Poetry and Drama)   
542     8845229041  Biblioteca Universale Rizzoli: Sulla Sponda De...   
...            ...                                                ...   
270794  014029953X                         Foe (Essential.penguin S.)   
270913  0340571187                                     Postmens House   
271094  8427201079                           El Misterio De Sittaford   
271182  0887781721                                          Tom Penny   
271196  3150013763                                     Der Hofmeister   

                   Book-Author  Year-Of-Publication  \
176                   Schiller                    0   
188     Gabriel Garcia Marquez                    0   
288         Barbara Kingsolver                    0   
351                    Golding                    0   
542                   P Coelho                    0   
...                        ...                  ...   
270794            J.M. Coetzee                    0   
270913        Maggie Hemingway                    0   
271094         Agatha Christie                    0   
271182             Tony German                    0   
271196              Jakob Lenz                    0   

                                 Publisher  \
176        Philipp Reclam, Jun Verlag GmbH   
188     Deutscher Taschenbuch Verlag (DTV)   
288                        Faber Faber Inc   
351        Fischer Taschenbuch Verlag GmbH   
542                     Fabbri - RCS Libri   
...                                    ...   
270794                   Penguin Books Ltd   
270913                    Trafalgar Square   
271094                    Editorial Molino   
271182                P. Martin Associates   
271196     Philipp Reclam, Jun Verlag GmbH   

                                              Image-URL-S  \
176     http://images.amazon.com/images/P/3150000335.0...   
188     http://images.amazon.com/images/P/342311360X.0...   
288     http://images.amazon.com/images/P/0571197639.0...   
351     http://images.amazon.com/images/P/3596214629.0...   
542     http://images.amazon.com/images/P/8845229041.0...   
...                                                   ...   
270794  http://images.amazon.com/images/P/014029953X.0...   
270913  http://images.amazon.com/images/P/0340571187.0...   
271094  http://images.amazon.com/images/P/8427201079.0...   
271182  http://images.amazon.com/images/P/0887781721.0...   
271196  http://images.amazon.com/images/P/3150013763.0...   

                                              Image-URL-M  \
176     http://images.amazon.com/images/P/3150000335.0...   
188     http://images.amazon.com/images/P/342311360X.0...   
288     http://images.amazon.com/images/P/0571197639.0...   
351     http://images.amazon.com/images/P/3596214629.0...   
542     http://images.amazon.com/images/P/8845229041.0...   
...                                                   ...   
270794  http://images.amazon.com/images/P/014029953X.0...   
270913  http://images.amazon.com/images/P/0340571187.0...   
271094  http://images.amazon.com/images/P/8427201079.0...   
271182  http://images.amazon.com/images/P/0887781721.0...   
271196  http://images.amazon.com/images/P/3150013763.0...   

                                              Image-URL-L  
176     http://images.amazon.com/images/P/3150000335.0...  
188     http://images.amazon.com/images/P/342311360X.0...  
288     http://images.amazon.com/images/P/0571197639.0...  
351     http://images.amazon.com/images/P/3596214629.0...  
542     http://images.amazon.com/images/P/8845229041.0...  
...                                                   ...  
270794  http://images.amazon.com/imag

we can see that there are 13 values which are greater than 20223 and 4618 values which are equal to zero in year of publication column which we have to treat before moving ahead.

In [ ]:
# Replacing Invalid years with max year
from collections import Counter
count = Counter(books['Year-Of-Publication'])
[k for k, v in count.items() if v == max(count.values())]

[2002]

In [ ]:
# lets replace all values of year of publicaion which are greater than 2023 and equal to zero with maximum number of years
books.loc[books['Year-Of-Publication'] > 2023, 'Year-Of-Publication'] = 2002
books.loc[books['Year-Of-Publication'] == 0, 'Year-Of-Publication'] = 2002

In [ ]:
# now lets verify if there any value greater than 2023
books.loc[books['Year-Of-Publication']>2022].count()

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
Image-URL-S            0
Image-URL-M            0
Image-URL-L            0
dtype: int64

In [ ]:
# Lets verfy if there and value equal to zero in year of publication column
books.loc[books['Year-Of-Publication']==0].count()

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
Image-URL-S            0
Image-URL-M            0
Image-URL-L            0
dtype: int64

In [ ]:
# checking null values
books.isnull().sum() 

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

there is one null value in Book-Author,Two null values in Publisher and three null values in Image-URL-L columns.Lets check what are those null values

In [ ]:
# checking null value in Book-Author column
books.loc[books['Book-Author'].isnull()]

ISBN                                         Book-Title  \
187689  9627982032  The Credit Suisse Guide to Managing Your Perso...   

       Book-Author  Year-Of-Publication                       Publisher  \
187689         NaN                 1995  Edinburgh Financial Publishing   

                                              Image-URL-S  \
187689  http://images.amazon.com/images/P/9627982032.0...   

                                              Image-URL-M  \
187689  http://images.amazon.com/images/P/9627982032.0...   

                                              Image-URL-L  
187689  http://images.amazon.com/images/P/9627982032.0...

there is one value in book author column which is NaN so lets replace it with 'Others'

In [ ]:
# lets replace nan value of book author with ohers 
books.at[187689 ,'Book-Author'] = 'Other'

In [ ]:
# lets check
books.loc[187689]

ISBN                                                          9627982032
Book-Title             The Credit Suisse Guide to Managing Your Perso...
Book-Author                                                        Other
Year-Of-Publication                                                 1995
Publisher                                 Edinburgh Financial Publishing
Image-URL-S            http://images.amazon.com/images/P/9627982032.0...
Image-URL-M            http://images.amazon.com/images/P/9627982032.0...
Image-URL-L            http://images.amazon.com/images/P/9627982032.0...
Name: 187689, dtype: object

In [ ]:
# We have replaced the null value with other
# checking null values in Publisher column
books.loc[books['Publisher'].isnull()]


ISBN       Book-Title      Book-Author  Year-Of-Publication  \
128890  193169656X      Tyrant Moon  Elaine Corvidae                 2002   
129037  1931696993  Finders Keepers  Linnea Sinclair                 2001   

       Publisher                                        Image-URL-S  \
128890       NaN  http://images.amazon.com/images/P/193169656X.0...   
129037       NaN  http://images.amazon.com/images/P/1931696993.0...   

                                              Image-URL-M  \
128890  http://images.amazon.com/images/P/193169656X.0...   
129037  http://images.amazon.com/images/P/1931696993.0...   

                                              Image-URL-L  
128890  http://images.amazon.com/images/P/193169656X.0...  
129037  http://images.amazon.com/images/P/1931696993.0...

In [ ]:
# Lets replace null values of Publisher column with others
books.at[128890,'Publisher']='Others'
books.at[129037,'Publisher']='Others'

In [ ]:
# lets verify for first
books.loc[128890]

ISBN                                                          193169656X
Book-Title                                                   Tyrant Moon
Book-Author                                              Elaine Corvidae
Year-Of-Publication                                                 2002
Publisher                                                         Others
Image-URL-S            http://images.amazon.com/images/P/193169656X.0...
Image-URL-M            http://images.amazon.com/images/P/193169656X.0...
Image-URL-L            http://images.amazon.com/images/P/193169656X.0...
Name: 128890, dtype: object

In [ ]:
# lets verify for second
books.loc[129037]

ISBN                                                          1931696993
Book-Title                                               Finders Keepers
Book-Author                                              Linnea Sinclair
Year-Of-Publication                                                 2001
Publisher                                                         Others
Image-URL-S            http://images.amazon.com/images/P/1931696993.0...
Image-URL-M            http://images.amazon.com/images/P/1931696993.0...
Image-URL-L            http://images.amazon.com/images/P/1931696993.0...
Name: 129037, dtype: object

In [ ]:
# we have transformed this values as well.Lets now proceecd for Image URL L column
# checking null value in Image-URL-L column
books.loc[books['Image-URL-L'].isnull()]

ISBN                                         Book-Title  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
220731  2070426769                 Peuple du cel,suivi de les Bergers   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

                  Book-Author  Year-Of-Publication          Publisher  \
209538     Michael Teitelbaum                 2000  DK Publishing Inc   
220731  Jean-Marie Gustave Le                 2003          Gallimard   
221678          James Buckley                 2000  DK Publishing Inc   

                                              Image-URL-S  \
209538  http://images.amazon.com/images/P/078946697X.0...   
220731  http://images.amazon.com/images/P/2070426769.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-M Image-URL-L  
209538  http://images.amazon.com/images/P/078946697X.0...         NaN  
220731  http://images.amazon.com/images/P/2070426769.0...         NaN  
221678  http://images.amazon.com/images/P/0789466953.0...         NaN

In [ ]:
# Since we will not be doing much with image url columns so lets drop it.
# Drop URL columns
books.drop(['Image-URL-S', 'Image-URL-M', 'Image-URL-L'], axis=1, inplace=True)
books.head()
     

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author  Year-Of-Publication                   Publisher  
0    Mark P. O. Morford                 2002     Oxford University Press  
1  Richard Bruce Wright                 2001       HarperFlamingo Canada  
2          Carlo D'Este                 1991             HarperPerennial  
3      Gina Bari Kolata                 1999        Farrar Straus Giroux  
4       E. J. W. Barber                 1999  W. W. Norton &amp; Company

In [ ]:
# Lets check is there any Duplicate values
books.duplicated().any()
     

False

In [ ]:
# lets check is there any null values
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271360 non-null  object
 3   Year-Of-Publication  271360 non-null  int64 
 4   Publisher            271360 non-null  object
dtypes: int64(1), object(4)
memory usage: 10.4+ MB
